AppCUI/src/Controls/NumericSelector.cpp

```diff
@@ -0,0 +1,176 @@
+    CREATE_TYPECONTROL_CONTEXT(NumericSelectorControlContext, Members, false);
+    Members->Layout.MinHeight = 1;
+    Members->Layout.MaxHeight = 1;
+    Members->Layout.MinWidth  = 10;
+    CHECK(Init(parent, "", layout, true), false, "Failed to create numeric selector!");
+    Members->Flags = GATTR_ENABLE | GATTR_VISIBLE | GATTR_TABSTOP;
+
+    Members->minValue = minValue;
+    Members->maxValue = maxValue;
+    Members->value    = value;
+
+    return true;
+}
+
+const unsigned long long NumericSelector::GetValue() const
+{
+    CREATE_TYPECONTROL_CONTEXT(NumericSelectorControlContext, Members, 0);
+    return Members->value;
+}
+
+const void NumericSelector::Enable()
+{
+    CREATE_TYPECONTROL_CONTEXT(NumericSelectorControlContext, Members, );
+    Members->Flags |= GATTR_ENABLE;
+}
+
+const void NumericSelector::Disable()
+{
+    CREATE_TYPECONTROL_CONTEXT(NumericSelectorControlContext, Members, );
+    Members->Flags ^= GATTR_ENABLE;
+}
+
+void NumericSelector::Paint(Renderer& renderer)
+{
+    CREATE_TYPECONTROL_CONTEXT(NumericSelectorControlContext, Members, );
+
+    const auto& textColor = [&]()
+    {
+        const auto& nsCfg = Members->Cfg->NumericSelector;
+
+        if (IsEnabled() == false)
+        {
+            return nsCfg.Inactive.TextColor;
+        }
+
+        if (Members->MouseIsOver)
+        {
+            return nsCfg.Hover.TextColor;
+        }
+
+        if (Members->Focused)
+        {
+            return nsCfg.Focused.TextColor;
+        }
+
+        return nsCfg.Normal.TextColor;
+    }();
+
+    renderer.WriteSingleLineText(0, 0, " - ", textColor);
+    renderer.DrawHorizontalLine(4, 0, Members->Layout.Width - 5, ' ', textColor);
+
+    // TODO: find another way!
+    LocalString<256> tmp;
+    tmp.Format("%llu", Members->value);
+
+    renderer.WriteSingleLineText(Members->Layout.Width / 2, 0, tmp.GetText(), textColor, TextAlignament::Center);
+    renderer.WriteSingleLineText(Members->Layout.Width - 3, 0, " + ", textColor);
+}
+
+bool NumericSelector::OnKeyEvent(Key keyCode, char AsciiCode)
+{
+    CREATE_TYPECONTROL_CONTEXT(NumericSelectorControlContext, Members, false);
+    switch (keyCode)
+    {
+    case Key::Left:
+        if (Members->minValue < Members->value)
+        {
+            Members->value--;
+            this->RaiseEvent(Event::EVENT_NUMERICSELECTOR_VALUE_CHANGED);
+        }
+        return true;
+
+    case Key::Right:
+        if (Members->maxValue > Members->value)
+        {
+            Members->value++;
+            this->RaiseEvent(Event::EVENT_NUMERICSELECTOR_VALUE_CHANGED);
+        }
+        return true;
+
+    default:
+        break;
+    }
+
+    return false;
+}
+
+void NumericSelector::OnMousePressed(int x, int y, MouseButton button)
+{
+    CREATE_TYPECONTROL_CONTEXT(NumericSelectorControlContext, Members, );
+
+    switch (button)
+    {
+    case MouseButton::Left:
+
+        // height is always 1 constrained - y doesn't matter
+
+        if (x < 4)
```

Comment:
Constrangerea asta hardcodwata cu width-ul 4 la butoane s-ar putea sa devină o problema mai târziu. Chestiile alea doua (+ și -) nu se pot face butoane cu context separat? Și sa primești evenimentele ca sunt pressed?

Reply:
da.
voi pune momentan paddingul pentru butoane intr o constanta urmand a extinde componenta printr-un (editable) field & doua butoane
https://github.com/gdt050579/AppCUI/projects/1#card-67686172

Reply:
Actually - tot nu rezolva problema (din contra butoanele sunt mai mari). Plus ca aceste controale ar cam trebui sa fie independente (daca schimbam ceva din logica de la Button) il poate impacta si pe acesta.
Pe de alta parte - limitarea de Width va exisa la toate contrroalele ==> daca e un Width mai mic deca o anumita valoare , nu se poate afisa ==> asa ca e mai bine sa il fixam noi (si framework-ul se asigura ca nu scazi sub valorea aia).

Reply:
mnda... si asta e corect

Reply:
Totusi nu as lasa '4' ==> l-as pune pune un constexpr sau macro (daca pe viitor vrem sa il facem mai mic / mare sa putem).

Reply:
Consider ca numeri selector, ca și componenta - este una complexa. Semantic este reprezentat din doua butoane și un label (sau text field). Atâta timp cât nu putem face componente complexe din componente atomice - e trist, pentru ca nu putem avea o creștere de abstractizare. Dacă se schimba butonul - se schimba și aici, true. Dar merge și invers, dacă se schimba butonul - nu se schimba și aici, iar semantic și logic se comporta ca doua butoane. 

Acum, dacă avem o limitare de width pentru un buton, trebuie sa o scoatem și sa punem una sensible, în care încape un caracter

Reply:
aici totodata am fi constransi la un min value la butoane (pe Y axis as well)
deci cumva toata componenta compusa ar trebui sa respecte aceleasi constrangeri
daca button Y min ar deveni prea mare pentru un context in care folosim numeric selector, ar cam distruge layoutul respectiv

Reply:
sa zicem ca.. daca cineva dorest, poate face un compound intre buttons +, - si un editable field drept o alta componenta entirely

Reply:
@beaverden e discutabil ce e control complex si ce nu este.  Daca e sa ne uitam la implementarile din alte framework-uri, un numeric selector e un control simplu. Acuma na - e ceva ce mai putem dezbate (uite - un list view pentru ca are capete de tabel care pana la urma se comporta ca niste butoane, ar trebui sa il consideram un control complex sau nu ?) Ca la ce implementari stiu eu, este un control mai degraba stand-alone.  Aici sunt deacord cu @gheorghitamutu ==> astea sunt controale de baza ca sa ai ceva de start ==> daca cineva vrea ceva super specializat isi poate deriva ceva din UserControl si sa il faca cum vrea el.

Reply:
Also - 2 butoane + 1 label cred ca va incurca functionalitate:
a) o sa mearga TAB-ul intre ele (e un pic cam ciudat la un Numeric Selector sa ai asa ceva)
b) nu o sa poti adauga extra stuff usor peste label (de exemplu - ar fi un nice feature sa poti sa scrii un numar si sa il inloucuiasca acolo ==> de genu daca tastez 1 apoi 2 apoi 3 ==> sa apara 123 ca si numar , respectiv daca tastez apoi BACK sa devina 12. Se poate pune un TextField ==> dar apoi nu o sa mai aiba un look & feel cu restul controalelor si va permite sa adaugi si litere & stuff. Idea e ca in final, s-ar putea sa ajungi sa scrii mai mult ca sa faci ceva 3 controale (tocmai ca sa limitetzi functionalitatea lor) vs sa ai ceva simplificat facut asa.